docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,293 @@
+# Flow Modules
+
+* Date: 2023-01-27
+* Author: Matt Durham @mattdurham
+* PR: [grafana/agent#2898](https://github.com/grafana/agent/pull/2898)
+* Status: Draft
+
+[Formatted Link for ease of user](https://github.com/grafana/agent/blob/rfc_modules/docs/rfcs/0007-flow-modules.md)
+
+## RFC Goals
+
+* Explain the use cases of modules
+* Explain what modules are
+* Go over possible syntax for modules
+* Go over pros and cons of modules
+
+## Summary
+
+One of the primary goals for the production usage of Agent Flow is parity with the static subsystem. One of the features of the static subsystem is [scraping service](https://github.com/grafana/agent/blob/main/docs/sources/configuration/scraping-service.md). Scraping service allows a user to run a fleet of agents and have thousands of scrape configurations distributed among running Agents. Through discussions within the Agent Team, we did not want to limit dynamically loading content to only scrape configs but allow most components to be loaded and used.
+
+During this time the Agent team saw a lot of potential in modules outside of scraping service. Packaging up sets of components for specific workflows, publishing common use cases and allowing better usage for internal customers in the Agent as a Service model.
+
+## Goals
+
+* Support single module loading via `module.single`
+* Support multiple module loading via `module.multiple`
+* Enable re-use of common patterns
+* Allow loading a module from a string
```

Comment:
Wanting to already shift this, I think it should be a label set with key ‘contents’ with the content and ‘id’ for an optional unique id. This will allow metadata to be passed since I also want to add a metadata component into the module.

Reply:
Torn over this for an ease of use mechanism. Maybe `module.single` loads from a string and `module.multiple` loads from a labelset.

Reply:
Aiming for consistency is something I think modules should value, since the mental model of how to use them shouldn't differ too wildly between loading a single module or many at a time. If that means an initial hurdle on getting familiar with the system (since modules are already somewhat of an intermediate feature in terms of difficulty) for the single use case, personally that feels like an acceptable tradeoff.